src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -784,8 +787,8 @@ public bool Remove(TKey key)
                         {
                             entries[last].next = entry.next;
                         }
-                        entry.hashCode = -1;
-                        entry.next = _freeList;
+
+                        entry.next = StartOfFreeList - _freeList;
```

Comment:
This can underflow in theory for very large `_freeList`. Add comments on why it does not happen?

Reply:
you mean because max hashtables len is [`0x7FEFFFFD`](https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Collections/HashHelpers.cs#L15)?

Reply:
Right. Assert that this does not underflow would be nice too.

Reply:
Added assertion where `StartOfFreeList` is used, let me know if it's enough(act as documentation) and feel free to suggest better message